layer4/routes.go

```diff
@@ -92,75 +104,70 @@ func (routes RouteList) Provision(ctx caddy.Context) error {
-			// a very lightweight copy of a pointer); this is important
-			// because this is a closure to the func below, which
-			// re-assigns the value as it compiles the middleware stack;
-			// if we don't make this copy, we'd affect the underlying
-			// pointer for all future request (yikes); we could
-			// alternatively solve this by moving the func below out of
-			// this closure and into a standalone package-level func,
-			// but I just thought this made more sense
-			nextCopy := next
-
-			// route must match at least one of the matcher sets
-			matched, err := route.matcherSets.AnyMatch(cx)
+func (routes RouteList) Compile(next Handler, logger *zap.Logger, matchingTimeout time.Duration) Handler {
+	return HandlerFunc(func(cx *Connection) error {
+		deadline := time.Now().Add(matchingTimeout)
+	router:
+		// timeout matching to protect against malicious or very slow clients
+		err := cx.Conn.SetReadDeadline(deadline)
+		if err != nil {
+			return err
+		}
+		for { // retry prefetching and matching routes until timeout
+			err = cx.prefetch()
 			if err != nil {
-				logger.Error("matching connection", zap.String("remote", cx.RemoteAddr().String()), zap.Error(err))
+				logFunc := logger.Error
+				if errors.Is(err, os.ErrDeadlineExceeded) {
+					err = ErrMatchingTimeout
+					logFunc = logger.Warn
+				}
+				logFunc("matching connection", zap.String("remote", cx.RemoteAddr().String()), zap.Error(err))
 				return nil // return nil so the error does not get logged again
 			}
-			if !matched {
-				return nextCopy.Handle(cx)
+			for _, route := range routes {
+				// A route must match at least one of the matcher sets
+				matched, err := route.matcherSets.AnyMatch(cx)
+				if errors.Is(err, ErrConsumedAllPrefetchedBytes) {
+					continue // ignore and try next route
+				}
+				if err != nil {
+					logger.Error("matching connection", zap.String("remote", cx.RemoteAddr().String()), zap.Error(err))
+					return nil
+				}
+				if matched {
+					// remove deadline after we matched
+					err = cx.Conn.SetReadDeadline(time.Time{})
+					if err != nil {
+						return err
+					}
+
+					lastHandler := HandlerFunc(func(conn *Connection) error {
+						// Catch potentially wrapped connection to use it as input for the next round of route matching.
+						// This is for example required for matchers after a tls handler.
+						cx = conn
+						// call the default/fallback handler
+						return next.Handle(conn)
+					})
+					var handler Handler
+					handler = &lastHandler
+					// compile the route handler stack with lastHandler being called last
+					for i := len(route.middleware) - 1; i >= 0; i-- {
+						handler = route.middleware[i](handler)
+					}
+					err = handler.Handle(cx)
+					if err != nil {
+						return err
+					}
+
+					// If matched route is terminal we stop routing,
+					// otherwise we jump back to the start of the routing loop to peel of more protocol layers.
+					if route.Terminal {
```

Comment:
Is there a way to do this without explicitly hard-coding whether a handler is terminal?

One problem I found when designing Caddy's HTTP routes and handlers was that, at first, I thought a handler was either terminal or not. For example, a file server was terminal because it loads the file from disk and writes the response. Nothing can come after it up the chain. An encoder module was never terminal, as it always encoded whatever the terminal handler up the chain would write.

But a cache handler was sometimes terminal and sometimes not. If it was a cache hit, it was terminal; if a miss, it would NOT be terminal because it would have to send the connection up the chain to be handled by the handler originating the content.

So one concern I have with this design is that it requires a handler to either be terminal or not, always.

Is there a way we can remove that `IsTerminal()` method and let it be determined implicitly by the flow of handling whether a handler is terminal? Even if it means putting a handler at the end that is a no-op and is always terminal.

Reply:
Yes I was also not a fan of IsTerminal. Now that you said it I think we can indeed get rid of it again. Because we already have a handler to catch wrapped connection (see [lastHandler](https://github.com/mholt/caddy-l4/pull/192/files#diff-ca91bc2c00b6e9985b0eeab47b53d0d3742328972c06b743ef18842f0f635cffR144)). I am sure we can also use it as terminal marker. Thanks for the idea I will try it tomorrow.

Reply:
@mholt I removed the IsTerminal function again. If a route is terminal or not is now determined by if the `lastHanlder` was called or not.

Reply:
Awesomeeee